# Masterthesis 
# Comparison of Functional Conservation 
# of Orthologs and Paralogs.

# Author:
## Chin-Hua Huang

# Supervisors:
## Prof. Dr. Kay Nieselt,
### Intergrative Transcriptomics in Wilhelm-Schickard Institute of University of Tübingen.
## Prof. Dr. Nadine Ziemert, 
### Interfaculty Institute of Microbiology and Infection Medicine, University of Tübingen.

## The website is still under modifications, but some of the content can already be viewed.
## After the approval of the supervisors , I will upload the thesis to this project.

## A Short Introduction of the topic of the thesis:
### My motivation comes from how we annotate the genes. There have been softwares developed for automatic gene annotations. Their algorithmic designs are based on assigning the gene functions of the orthologs to the unknown genes. Many researchers casted doubt on this approach that whether the assignment of gene function of an orthologous gene are always a optimal strategy to annotate a gene. In 2011, Nehrt et al. coined the applied  functional similarity analysis on GO-terms and expression correlation analysis on microarray datasets.  After the refutation of the conjecture by Nehrt et al., there were many articles that examined ortholog conjecture. In 2020, Stamboulian et al. examined once more the ortholog conjecture with functional similarity analyses and showed once again that the ortholog conjecture does not hold on the gene annotations in the UniProt database. In this thesis, the works of Stamboulian et al. (2020) are examined and extended, including more species pairs and the expression correlation analyses on RNA-seq datasets to view this issue in a more broader context.

### For the literatur about this topic, please refer to:
#### 1. Moses Stamboulian, Rafael F. Guerrero, Matthew W. Hahn, and Predrag Radivojac. The ortholog conjecture revisited: the value of orthologs and paralogs in function prediction. Bioinformatics, 36:219–226, 2020.

#### 2. Xiaoshu Chen and Jianzhi Zhang. The ortholog conjecture is untestable by the current gene ontology but is supported by RNA sequencing data. PLOS Computational Biology, 8(11), 2012.

#### 3. Adrian M. Altenhoff, Romain A. Studer, Marc RobinsonRechavi, and Christophe Dessimoz. Resolving the ortholog conjecture: Orthologs tend to be weakly, but significantly, more similar in function than paralogs. PLOS Computational Biology, 8(5), 2012.

#### 4. Nathan L. Nehrt, Wyatt T. Clark, Predrag Radivojac, and Matthew W. Hahn. Testing the ortholog conjecture with comparative functional genomic data from mammals. Bioinformatics, 7(6), 2011.

## The Introduction of the Directories:
### Directory codes_and_files includes the raw data, the pre-processed data and the implementations of the functional similarity and expression correlation analyses.
### Directory figures includes the plots that are generated for the comparison of functional similarity and expression correlation of every species pair. 

### There are more README files for detailed introductions in each directory.
