# List of files in the directory gos:
## all_protein_seq_xx.fasta : The proteomes of the species pair Saccharomyces cerevisiae and Schizosaccharomyces pombe (_cp.fasta), species pair human and mouse (_hm.fasta) and species pair mouse and rat (_mr.fasta) generated by OrthoVenn2.

## ortho_co_orthologs_xx.txt, ortho_orthologs_xx.txt and ortho_inparalogs_xx.txt: The files of co-orthologs, orthologs and inparalogs predicted by OrthoVenn2 for the species pairs Saccharomyces cerevisiae and Schizosaccharomyces pombe (_cp.txt), species pair human and mouse (_hm.txt) and species pair mouse and rat (_mr.txt). 

## bayes_prob_table_xx.tsv : The tables for the reference of conditional probability for biological process (_bp.tsv), cellular component (_cc.tsv) and molecular function (_mf.tsv). 
###   &nbsp  &nbsp &nbsp &nbsp There are 4 main columns: <br />
###      &nbsp &nbsp &nbsp                                                  child_node: the GO-term itself <br />
###      &nbsp &nbsp &nbsp                                                  joint_count: the occurrences that all of the parents are all in the same gene annotations. <br />
###      &nbsp &nbsp &nbsp                                                  joint_count_child: the occurrences that all of the parents and the GO-term itself are all in the same gene  annotations. <br />
###      &nbsp &nbsp &nbsp                                                  cond_prob: the conditional probability given that all of the parents are all in the same gene annotations, 
###                                                                   what is the probability that this GO-term itself also is the same gene annotations. <br /> 

### child_parent_xx.tsv : The tables for the child-parent relationship of the GO-terms for biological process (_bp.tsv), cellular component (_cc.tsv) and molecular function (_mf.tsv).
###                          There are two columns: The first column is somehow marked with 2, is the GO-term itself.
###                                                 The second column is the parents of the GO-term concatenated with the connector ",".

### expr_go.py : The program for the evaluation of the expression correlation w.r.t. functional similarity. The code to excecute the program is in master.sh, one directory up.

### go.obo : The original file that embeds the gene ontology files, can be obtained under the website: http://purl.obolibrary.org/obo/go.obo
### go_counts_xx.tsv : The occurrence of every GO-term. There are 2 columns: The first column : GO-term itself.
###                                                                             The second column is somehow marked with 0: The occurrence of this GO-term.

### xxx_uniprot.tsv: They are the files that map Ensembl to UniProt accession number. 
###                     There are the original mapping files of human (human_), mouse (mouse_), rat (rat_), S. cerevisiae (cerev_) and S. pombe (pombe_).
###                     They can be obtained from the following websites:
###                     S. cerevisiae : http://ftp.ebi.ac.uk/ensemblgenomes/pub/release-51/fungi/tsv/saccharomyces_cerevisiae/
###                     S. pombe : http://ftp.ebi.ac.uk/ensemblgenomes/pub/release-51/fungi/tsv/schizosaccharomyces_pombe/
###                     Human : http://ftp.ensembl.org/pub/release-103/tsv/homo_sapiens/
###                     mouse : http://ftp.ensembl.org/pub/release-103/tsv/mus_musculus/
###                     rat : http://ftp.ensembl.org/pub/release-103/tsv/rattus_norvegicus/

### semsim.py: The program for the computation of functional similarity measures. The code to excecute the program is in master.sh, one directory up.
###  
