# List of files in the directory gos:
## all_protein_seq_xx.fasta : 
### The proteomes of the species pair Saccharomyces cerevisiae and Schizosaccharomyces pombe (_cp.fasta), human and mouse (_hm.fasta) and species pair mouse and rat (_mr.fasta) generated by OrthoVenn2.

## ortho_co_orthologs_xx.txt, ortho_orthologs_xx.txt and ortho_inparalogs_xx.txt: 
### The files of co-orthologs, orthologs and inparalogs predicted by OrthoVenn2 for the species pairs Saccharomyces cerevisiae and Schizosaccharomyces pombe (_cp.txt), human and mouse (_hm.txt) and species pair mouse and rat (_mr.txt). 

## go.obo : 
### The original file that embeds the gene ontology files, can be obtained under the website: http://purl.obolibrary.org/obo/go.obo

## xxx_uniprot.tsv: 
### They are the files that map Ensembl to UniProt accession number. 
### There are the original mapping files of human (human_), mouse (mouse_), rat (rat_), S. cerevisiae (cerev_) and S. pombe (pombe_). 
###  They can be obtained from the following websites: 
####   &nbsp; &nbsp; &nbsp; S. cerevisiae :http://ftp.ebi.ac.uk/ensemblgenomes/pub/release-51/fungi/tsv/saccharomyces_cerevisiae/Saccharomyces_cerevisiae.R64-1-1.51.uniprot.tsv.gz 
####   &nbsp; &nbsp; &nbsp; S. pombe : http://ftp.ebi.ac.uk/ensemblgenomes/pub/release-51/fungi/tsv/schizosaccharomyces_pombe/Schizosaccharomyces_pombe.ASM294v2.51.uniprot.tsv.gz 
####   &nbsp; &nbsp; &nbsp; Human : http://ftp.ensembl.org/pub/release-103/tsv/homo_sapiens/Homo_sapiens.GRCh38.103.uniprot.tsv.gz 
####   &nbsp; &nbsp; &nbsp; Mouse : http://ftp.ensembl.org/pub/release-103/tsv/mus_musculus/Mus_musculus.GRCm39.103.uniprot.tsv.gz 
####   &nbsp; &nbsp; &nbsp; Rat : http://ftp.ensembl.org/pub/release-103/tsv/rattus_norvegicus/Rattus_norvegicus.Rnor_6.0.103.uniprot.tsv.gz
### To use the lastest release, please reduce the links to /pub/.. and navigate to the latest release.

## semsim.py: 
### The program for the computation of functional similarity measures. The code to excecute the program is in master.sh, one directory up.
### The program generates the summary_gos_xx_xx.tsv for hm (human and mouse), mr (mouse and rat) and cp (S. cerevisiae and S. pombe) for biological process (_bp.tsv), cellular component (_cc.tsv) and molecular function (_mf.tsv).
### The summary_gos_xx_xx.tsv contains the UniProt accessions of the gene pairs, the sequences and the resulting sequence identity (seq_iden) and functional similarity (yc_sim: Yang-Clark similarity; sc_sim: Schlicker similarity).

## expr_go.py : 
### The program for the evaluation of the expression correlation w.r.t. functional similarity. The code to excecute the program is in master.sh, one directory up.
### The program generates the summary_expr_go_xx_xx.tsv which contains the information of expression correlations (r_sim: rank-score similarity, p_corr: Pearson correlation coefficient.) and functional similarity (yc_sim: Yang-Clark similarity; sc_sim: Schlicker similarity) of the species pair human and mouse (human_mouse) and the comparative mouse and rat RNA-seq datasets of SÃ¶llner et al. (mouse_rat_soellner) and Fushan et al. (mouse_rat_fushan) for biological process (_bp.tsv), cellular component (_cc.tsv) and molecular function (_mf.tsv).

## prob_go.py: 
### The program for the computation of the occurrences of GO-terms, conditional probability of child and parent GO-terms and the relative frequency of each GO-term w.r.t. their respective root term. The code to excecute the program is in master.sh, one directory up. 

### This program takes the following input files:
#### go.obo : 
##### The file that embeds the 3 gene ontology graphs, can be obtained under the website: http://purl.obolibrary.org/obo/go.obo

#### goa_uniprot_all.gaf.gz:
##### This file is absent in the directory because it is too big for GitHub.
##### It contains all gene annotations in the UniProt database.
##### Please download the file in the following address:
##### http://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/goa_uniprot_all.gaf.gz

### This program generates the following output files:

#### child_parent_xx.tsv : 
##### The tables for the child-parent relationship of the GO-terms for biological process (_bp.tsv), cellular component (_cc.tsv) and molecular function (_mf.tsv).
##### There are two columns: 
###### 1. column is somehow marked with 2, is the GO-term itself. 
###### 2. column is the parents of the GO-term concatenated with the connector ",".

#### paths_go_xx.tsv: 
##### The paths from a GO-term to its root node for biological process (_bp.tsv), cellular component (_cc.tsv) and molecular function (_mf.tsv).
##### This file contains 2 columns:
###### 1. column: The GO-term itself.
###### 2. column: The paths from this GO-term to its root term, concatenated with the connector ",".

#### paths_prob_xx.tsv:
##### The paths covered by the "propagated annotations" (Please refer to the Materials and Methods section of my thesis.) of a gene in UniProt for biological process (_bp.tsv), cellular component (_cc.tsv) and molecular function (_mf.tsv).
##### This file contains 4 columns:
###### 1. column: Index.
###### id: The UniProt accession of a gene.
###### go_term: The GO-terms that are assigned to this gene, concatenated with the connector ",".
###### path: The "propagated annotations" to this gene, concatenated with the connector ",".

#### go_counts_xx.tsv : 
##### The occurrence of every GO-term for biological process (_bp.tsv), cellular component (_cc.tsv) and molecular function (_mf.tsv). 
##### There are 2 columns: 
###### 1. column : GO-term itself. 
###### 2. column is somehow marked with 0: The occurrence of this GO-term.


#### bayes_prob_table_xx.tsv : 
##### The tables for the reference of conditional probability for biological process (_bp.tsv), cellular component (_cc.tsv) and molecular function (_mf.tsv). 
##### There are 5 columns: 
###### 1. column: Index
###### child_node: The GO-term itself. 
###### joint_count: The occurrences that all of the parents are all in the same gene annotations. 
###### joint_count_child: The occurrences that all of the parents and the GO-term itself are all in the same gene annotations. <br />
###### cond_prob: The conditional probability given that all of the parents are all in the same gene annotations, the probability that this GO-term itself also is in the same gene annotations. 

#### probs_xx.tsv : 
##### The tables that consist of the columns of the TSV-files about GO-terms listed above for biological process (_bp.tsv), cellular component (_cc.tsv) and molecular function (_mf.tsv). 
##### There are 5 columns: 
###### 1. column: Index
###### node: The GO-term itself. 
###### path: The path from the GO-term to the root node.
###### joint_count: The occurrences that all of the parents are all in the same gene annotations. 
###### joint_count_child: The occurrences that all of the parents and the GO-term itself are all in the same gene annotations. <br />
###### cond_prob: The conditional probability given that all of the parents are all in the same gene annotations, the probability that this GO-term itself also is in the same gene annotations.
###### count_ancestor: The count of the root term.
###### rel_freq: The count of the GO-term w.r.t. the root term.







